static/app/components/tooltip.tsx

```diff
@@ -320,67 +240,56 @@ function Tooltip({
-                duration: 0.2,
-              }),
-            }}
-            exit={{
-              opacity: 0,
-              scale: 0.95,
-              transition: testableTransition({type: 'spring', delay: 0.1}),
-            }}
-            style={computeOriginFromArrow(position, arrowProps)}
-            transition={{duration: 0.2}}
-            className="tooltip-content"
-            aria-hidden={!visible}
-            ref={ref}
-            data-placement={placement}
-            popperStyle={popperStyle}
-            onMouseEnter={() => isHoverable && handleOpen()}
-            onMouseLeave={() => isHoverable && handleClose()}
-          >
-            {title}
-            <TooltipArrow
-              ref={arrowProps.ref}
-              data-placement={placement}
-              style={arrowProps.style}
-            />
-          </TooltipContent>
-        </PositionWrapper>
-      )}
-    </Popper>
-  ) : null;
-
   return (
     <Manager>
       <Reference>{({ref}) => renderTrigger(children, ref)}</Reference>
-      {createPortal(<AnimatePresence>{tip}</AnimatePresence>, tooltipPortal)}
+      {createPortal(
+        <AnimatePresence>
+          {forceShow || isOpen ? (
+            <Popper placement={position} modifiers={modifiers}>
+              {({ref, style, placement, arrowProps}) => (
+                <PositionWrapper style={style}>
+                  <TooltipContent
+                    id={tooltipId}
+                    initial={{opacity: 0}}
+                    animate={{
+                      opacity: 1,
+                      scale: 1,
+                      transition: testableTransition({
+                        type: 'linear',
+                        ease: [0.5, 1, 0.89, 1],
+                        duration: 0.2,
+                      }),
+                    }}
+                    exit={{
+                      opacity: 0,
+                      scale: 0.95,
+                      transition: testableTransition({type: 'spring', delay: 0.1}),
+                    }}
+                    style={computeOriginFromArrow(position, arrowProps)}
+                    transition={{duration: 0.2}}
+                    className="tooltip-content"
+                    aria-hidden={false}
+                    ref={ref}
+                    data-placement={placement}
+                    popperStyle={popperStyle}
+                    onMouseEnter={() => isHoverable && handleOpen()}
+                    onMouseLeave={() => isHoverable && handleClose()}
+                  >
+                    {title}
+                    <TooltipArrow
+                      ref={arrowProps.ref}
+                      data-placement={placement}
+                      style={arrowProps.style}
+                    />
+                  </TooltipContent>
+                </PositionWrapper>
+              )}
+            </Popper>
+          ) : null}
+        </AnimatePresence>,
+        document.body
+      )}
```

Comment:
I'm kind of 👎 to nesting everything like this, any reason you can't keep it with the `const tip` so we have less indentation, making things just a bit easier to mentally keep track of?

Reply:
I would actually say the reverse. Keeping things in a single return makes it easier to read and follow code flow. In other words, I find it more straightforward. I've seen this pattern in a lot of our components and I dont like it because we are manually trying to adjust for the fact that we have a long and possibly too complex of a component.

To put it as an example, assume some long component like
```tsx
function Component() {
  const ComponentToRender = someFlag || maybeSomeOtherFlag ? (
     <div>
      s
      o
      m
      e

      l
      o
      n
      g
    
      b
      o
      d
      y
     </div>
  ) : null
  
  return (
    <Tooltip>
     ... more tooltip stuff
     {ComponentToRender}
    </Tooltip>
  )
}
```
If you read above (and assume that there may be things between const ComponentToRender and the final return so it may be even worse). If you are reading the return/render and want to know what ComponentToRender means or when will ComponentToRender be visible, you have to either remember the `someFlag || maybeSomeOtherFlag` expression and keep it in memory, or go to definition and then come back to where you were. This is imo an example of indirection that makes it harder for new people to onboard - it's hard to remember all this information when you are trying to understand what is happening

Now if we have a look at a different example:
```
function Component() {
  return (
    <Tooltip>
     ... more tooltip stuff
     {someFlag || maybeSomeOtherFlag ? (
        <div>
         s
         o
         m
         e

         l
         o
         n
         g
    
         b
         o
         d
         y
        </div>
     ) : null}
    </Tooltip>
  )
}
```
Now if you are reading ^ and want to know when will the "some long body" fn component render, you dont need to jump around the file, it's right there. The problem you have now is that maybe your long body is doing too much, so it may make sense to take the long fn body and separate it. 

but ^ example is still not the best, as the long body is annoying to read. What you would ideally want is something more like this

```ts
function Component() {
  return (
    <Tooltip>
     ... more tooltip stuff
     {someFlag || maybeSomeOtherFlag ? (<ComponentWithLongBody>) : null}
    </Tooltip>
  )
}
```

To be fair, because in our case it seems that most of the body is taken by the animate props. Those are static, so we can just abstract them away, maybe define them at top of the file. They are generally something that we will not be touching often, so that may already help a lot

Reply:
+1 to abstracting away prop constants for clarity if we're worried about length/readability of the return here.

personally i'm not a huge fan of ternarys to determine component rendering.
I'd either opt for a `{someFlag || maybeSomeOtherFlag && (<Component />)}`}
And if we want to render something else on the falsey, just do a completely new block
```
...
{
  maybeTruthy && (
    <ComponentThing />
  )
}
{
  !maybeTruthy && (
    <OtherComponentThing />
  )
}
...
```
I can see this being worst of both worlds too tho, but to me it's infintely easier to understand than trying to find a `:` amongst all the component props.

Alternatively are you all opposed to having sub render methods?
I'm also personally partial to breaking out pieces into subrender methods rather than constants

ie. 
```
const Page = () => {

  const renderHeader = () => {
    if (maybeTruthy) {
      return (<ComponentThing />)
    }
    return <OtherComponentThing />
  }

  return (
    <div>
      {renderHeader()}
      <Foo />
    </div>
  )
}
```

🤷 Let me know if you guys hate this suggestion.
Just what i've been used to doing

Reply:
actually i guess having subrenders is more a class pattern than a functional component pattern 🤔 

in which case, i think i'm fine with breaking it out into a separate component or leaving it in line rather than potentially having multiple component constants floating around with other variables nestled in between

Reply:
> actually i guess having subrenders is more a class pattern than a functional component pattern 

Right, that's why I had it rendering all the interesting tooltip stuff in the `tip` variable.

My argument for this btw is basically that I want to keep what's "interesting" separate from what's "not interesting".

To me the `Manager`, `AnimatePresence`, and `createPortal` is all a separate concern from the actual rendering of the tooltip, which is why i separated it out. If it's all rendered in one big, pyramid shaped, render, it's IMO visually harder to keep track of what I'm looking at.

Reply:
> +1 to abstracting away prop constants for clarity if we're worried about length/readability of the return here.
I dont think we are? Like if I look at the component return, I can easily see the whole block on my 14" screen + I have a larger font. It's ~40loc, I dont think that warrants breaking this down tbh.

`{renderHeader()}` is an antipattern for functional react code - you should create a PageHeader component there (stable component reference that wont waste render calls)

Another thing to note is that `{someFlag || maybeSomeOtherFlag && (<Component />)}` is not the same comparison to `const Component = someFlag || maybeSomeOtherFlag && <Component />` because Component here can be false/null which is hidden when you use it down the line - hence why you need interpolation (this is the main reason why I'm arguing this)
```tsx
<div>
{Component} // <- Do you know if this actually renders a component?
</div>
```

Reply:
It wouldn’t be capitalized though in that case by convention right?

Reply:
>  I can easily see the whole block on my 14" screen + I have a larger font. It's ~40loc, I dont think that warrants breaking this down tbh

🤷 yea - if it was like 100+ lines of component, id be down to split it up into separate components or break out prop constants. If it's not that, then personally i'm fine with keeping it in line

corrected myself on the functional anti-pattern too

as for the ternary alternative using `&&` my issue/solution is more with the use of ternary, not with inline vs a constant. I agree, i'd rather have it interpolated in line than splitting it out into a component (I would only suggest splitting it out in a class component as a subRender)

for this case i mentioned i agree with either keeping it all inline in the return OR creating separate functional components - but not breaking it out into variable components

Reply:
> It wouldn’t be capitalized though in that case by convention right?

Yeah, my thought on this is that the value can represent actual component that's rendering, then we should capitalize it so that the variable reflects that. When it's a simple value we are interpolating, I like to keep it lowercase.

Reply:
i'm late to the party hehe but thought I'd share my thoughts anyway.
I agree with @JonasBa on this:

> If you are reading the return/render and want to know what ComponentToRender means or when will ComponentToRender be visible, you have to either remember the someFlag || maybeSomeOtherFlag expression and keep it in memory, or go to definition and then come back to where you were. This is imo an example of indirection that makes it harder for new people to onboard - it's hard to remember all this information when you are trying to understand what is happening

I find myself doing ^ a lot and it adds to the cognitive load when reading new component files.

By encapsulating components and their logic in a variable like `tip`, we're masking the actual complexity of the component. **IMO if there's a need to break down render into digestible chunks then maybe it's a sign that the component is doing too much and or that it's too complicated and needs to be simplified**